web-client/extras/lib/BufferUtils.ts

```diff
@@ -0,0 +1,363 @@
+                ((c & 0xFC00) === 0xD800) && (i + 1) < str.length
+                && ((str.charCodeAt(i + 1) & 0xFC00) === 0xDC00)
+            ) {
+                // Surrogate Pair
+                c = 0x10000 + ((c & 0x03FF) << 10) + (str.charCodeAt(++i) & 0x03FF);
+                out[p++] = (c >> 18) | 240;
+                out[p++] = ((c >> 12) & 63) | 128;
+                out[p++] = ((c >> 6) & 63) | 128;
+                out[p++] = (c & 63) | 128;
+            } else {
+                out[p++] = (c >> 12) | 224;
+                out[p++] = ((c >> 6) & 63) | 128;
+                out[p++] = (c & 63) | 128;
+            }
+        }
+        return new Uint8Array(out);
+    }
+
+    private static _utf8TextEncoder(str: string): Uint8Array {
+        if (typeof TextEncoder === 'undefined') throw new Error('TextEncoder not supported');
+        if (BufferUtils._UTF8_ENCODER === null) throw new Error('TextEncoder does not support utf8');
+        if (BufferUtils._UTF8_ENCODER === undefined) {
+            try {
+                BufferUtils._UTF8_ENCODER = new TextEncoder();
+            } catch (e) {
+                BufferUtils._UTF8_ENCODER = null;
+                throw new Error('TextEncoder does not support utf8');
+            }
+        }
+        return BufferUtils._UTF8_ENCODER.encode(str);
+    }
+
+    static fromUtf8(str: string): Uint8Array {
+        if (typeof TextEncoder !== 'undefined' && BufferUtils._UTF8_ENCODER !== null) {
+            try {
+                return BufferUtils._utf8TextEncoder(str);
+            } catch (e) {
+                // Disabled itself
+            }
+        }
+        return BufferUtils._strToUint8Array(str);
+    }
+
+    static fromAny(o: Uint8Array | string, length?: number): SerialBuffer {
+        if (o === '') return SerialBuffer.EMPTY;
+        if (!o) throw new Error('Invalid buffer format');
+        if (o instanceof Uint8Array) return new SerialBuffer(o);
+        try {
+            return BufferUtils.fromHex(o, length);
+        } catch (e) {
+            // Ignore
+        }
+        try {
+            return BufferUtils.fromBase64(o, length);
+        } catch (e) {
+            // Ignore
+        }
+        throw new Error('Invalid buffer format');
+    }
+
+    static concatTypedArrays<T extends TypedArray>(a: T, b: T): T;
+    static concatTypedArrays(a: any, b: any): TypedArray {
+        const c = new (a.constructor)(a.length + b.length);
+        c.set(a, 0);
+        c.set(b, a.length);
+        return c;
+    }
+
+    static equals(a: TypedArray, b: TypedArray): boolean {
+        const viewA = BufferUtils._toUint8View(a);
+        const viewB = BufferUtils._toUint8View(b);
+        if (viewA.length !== viewB.length) return false;
+        for (let i = 0; i < viewA.length; i++) {
+            if (viewA[i] !== viewB[i]) return false;
+        }
+        return true;
+    }
+
+    /**
+     * Returns -1 if a is smaller than b, 1 if a is larger than b, 0 if a equals b.
```

Comment:
Should probably add: Shorter arrays are always considered smaller than longer ones.

Reply:
Addressed in #2975